### PR TITLE
Clean up static functions in staking protocol

### LINKED ISSUE #2142: Clean up some static functions in staking protocol
**What would you like to be added**:
We have some static functions like getTotalBucketCount (not used anywhere), getBucket, updateBucket, putBucket, delBucket, getAllBuckets, getBucketsWithIndices, putBucketAndIndex, which could all be moved in to CandidateStateManager. Also, function settleAction could be moved into CandidateStateManager as well.

**Why is this needed**:
Because of these functions, we have a mixture usage of CandidateStateManager and StateManager throughout this package. By moving them into CandidateStateManager, we don't need to expose StateManager APIs, and CandidateStateManager could be the only interface interacting with this protocol internally.

### PR DESCRIPTION
**Pull request by newyear6565**
Closes #2142

### CODE REVIEW

File: action/protocol/staking/candidate_statemanager.go [hunk 1]
```diff
@@ -173,3 +180,129 @@ func (csm *candSM) Commit() error {
 	// write update view back to state factory
 	return csm.WriteView(protocolID, ConvertToViewData(csm))
 }
+
+func (csm *candSM) updateBucket(index uint64, bucket *VoteBucket) error {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_statemanager.go:184]: think these private methods are to be added to CandidateStateManager interface
so protocol.StateManager can be removed

Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:184]: according to  #2142 description

Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:184]: also, latest code has changed quite a bit, pls rebase the PR

File: action/protocol/staking/candidate_statemanager.go [hunk 2]
```diff
@@ -22,9 +27,28 @@ const (
...
+		delBucketIndex(key []byte, index uint64) error
+	}
+	// CandidateSet related to setting candidates
+	CandidateSet interface {
+		delCandidate(name address.Address) error
+		putCandidate(d *Candidate) error
+	}
 	// CandidateStateManager is candidate state manager on top of StateManager
 	CandidateStateManager interface {
 		protocol.StateManager
```
Comment (dustinxie) [inline action/protocol/staking/candidate_statemanager.go:47]: #2142 wants CandidateStateManager not to expose StateManager interface 
so we need to implement methods like PutState(), DelState() for (*candSM)
then we can delete "protocol.StateManager" here
@CoderZhi  can confirm

Comment (dustinxie) [reply action/protocol/staking/candidate_statemanager.go:47]: should look like:
candSM struct {
		sm protocol.StateManager
		candCenter *CandidateCenter
		bucketPool *BucketPool
	}

and 
func (csm *candSM) PutState(s interface{}, options ...StateOption) (uint64, error) {
    return csm.sm.PutState(s, options...)
}

Comment (newyear6565) [reply action/protocol/staking/candidate_statemanager.go]: After hide `StateManager ` interface , I think we may not modify `candSM` structure. If `candSM` is not changed , it has `StateManager` interface , so there is no need to implement functions like `PutState()`, `DelState()` for `candSM`.The code will be more simplified.

Comment (newyear6565) [reply action/protocol/staking/candidate_statemanager.go]: just like `candSR` in candidate_statereader.go

File: action/protocol/staking/candidate_statemanager.go [hunk 3]
```diff
@@ -128,26 +162,193 @@ func (csm *candSM) Upsert(d *Candidate) error {
...
+	return err
+}
+
+func (csm *candSM) delCandidate(name address.Address) error {
+	_, err := csm.DelState(protocol.NamespaceOption(CandidateNameSpace), protocol.KeyOption(name.Bytes()))
+	return err
+}
+
+// settleAccount deposits gas fee and updates caller's nonce
+func (csm *candSM) settleAction(
```
Comment (dustinxie) [inline action/protocol/staking/candidate_statemanager.go:320]: this should be staking protocol's func, not csm's
it's kinda counter-intuitive that csm needs to take p *Protocol as input

Comment (newyear6565) [reply action/protocol/staking/candidate_statemanager.go:320]: ok , i will revert it

Comment (newyear6565) [reply action/protocol/staking/candidate_statemanager.go]: reverted @dustinxie

**Pull request closed without merging**